# Fake News Detector



The goal of this project is to learn methods of categorizing data by identifying useful attributes from our data in order to fit the data to a model. Specifically, we will be using the following numerical statistic to fit a Passive-Aggressive classification model:

**Term Frequency (TF)** : The number of times a term occurs in a document. A higher value means a term occurs more often than others.

**Inverse Document Frequency (IDF)**: Measure of how irrelevant a term is if it occurs too often across many other documents.


Using TF will allow our Passive-Aggressive Classifier to passively learn to classify articles based on occurrences of specific keywords to determine real news. The IDF will allow the model to make corrections in future predictions with further classification of specific keyterms with negative feedback between different news articles.

1. Import necessary packages from numpy, pandas, and sklearn.

In [3]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

2. Read the data into a Dataframe, get the shape of the data and the first 5 records to get a feel of the data we are working with.

In [15]:
dataframe = pd.read_csv('./news.csv')

dataframe.shape
dataframe.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

3. Get labels from the data

In [ ]:
labels = dataframe.label
labels.head()

4. Separate training data from testing data.

In [7]:
x_train, x_test, y_train, y_test = train_test_split(dataframe['text'], labels, test_size=0.2, random_state=7)

5. Initialize a TfidfVectorizer with English stop words with a document frequency of 0.7 to process our data set into a matrix of TF-IDF features. Fit and transform the vectorizer on the train set and transform the test set.


In [8]:
tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7)

tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test = tfidf_vectorizer.transform(x_test)

6. Initialize a PassiveAggressiveClassifier and fit this on tfidf_train and y_train. Then, predict on the test set from the TfidfVectorizer and calculate the accuracy using accuracy_score().

In [9]:
pac = PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train, y_train)

y_pred = pac.predict(tfidf_test)
score = accuracy_score(y_test, y_pred)
print(f'Accuracy: {round(score*100, 2)}%')

Accuracy: 93.21%


7. Print out a confusion matrix to gain insight into number of true and false negatives and positives.

In [11]:
confusion_matrix(y_test, y_pred, labels=['FAKE', 'REAL'])

array([[594,  44],
       [ 42, 587]])

Our model was able to identify 594 true positives, 587 true negatives, 42 false positives, and 44 false negatives.

---

*This is a project intended for educational purposes. Project guided by* [Data Flair](https://data-flair.training/blogs/advanced-python-project-detecting-fake-news/)